File: app/src/main/java/nl/acidcats/tumblrlikes/ui/screens/photo_screen/PhotoFragment.java
package nl.acidcats.tumblrlikes.ui.screens.photo_screen;

import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.target.DrawableImageViewTarget;
import com.bumptech.glide.request.target.Target;

import javax.inject.Inject;

import butterknife.BindView;
import nl.acidcats.tumblrlikes.R;
import nl.acidcats.tumblrlikes.core.constants.FilterType;
import nl.acidcats.tumblrlikes.core.usecases.photos.PhotoViewUseCase;
import nl.acidcats.tumblrlikes.di.AppComponent;
import nl.acidcats.tumblrlikes.ui.screens.base.BaseFragment;
import nl.acidcats.tumblrlikes.ui.screens.photo_screen.viewmodels.PhotoActionDialogViewModel;
import nl.acidcats.tumblrlikes.ui.screens.photo_screen.widgets.InteractiveImageView;
import nl.acidcats.tumblrlikes.ui.screens.photo_screen.widgets.PhotoActionDialog;
import nl.acidcats.tumblrlikes.ui.screens.photo_screen.widgets.PhotoNavBar;
import nl.acidcats.tumblrlikes.util.GlideApp;

/**
 * Created by stephan on 13/04/2017.
 */

public class PhotoFragment extends BaseFragment implements PhotoScreenContract.View {
    private static final String TAG = PhotoFragment.class.getSimpleName();

    private static final long HIDE_UI_DELAY_MS = 2000L;

    @Inject
    PhotoViewUseCase _photoViewUseCase;
    @Inject
    PhotoScreenPresenter _presenter;

    @BindView(R.id.photo)
    InteractiveImageView _photoView;
    @BindView(R.id.photo_action_dialog)
    PhotoActionDialog _photoActionDialog;
    @BindView(R.id.photo_nav_bar)
    PhotoNavBar _photoNavBar;

    private Handler _handler = new Handler();
    private Runnable _uiHider = this::hideUI;
    private boolean _isTest = true;

    public static PhotoFragment newInstance() {
        return new PhotoFragment();
    }

    @Override
    protected void injectFrom(AppComponent appComponent) {
        appComponent.inject(this);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        _presenter.restoreState(savedInstanceState, getArguments());
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_photo, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        _presenter.setView(this);
        _presenter.onViewCreated();

        _photoView.setGestureListener(this::onGesture);

        _photoActionDialog.setPhotoActionListener(_presenter);

        _photoNavBar.setFilterOptionSelectionListener(filterType -> _presenter.onFilterTypeSelected(filterType));

        if (_isTest) {
            _photoView.setAlpha(0.1f);
        }
    }

    private void onGesture(InteractiveImageView.Gesture gesture) {
        switch (gesture) {
            case SIDE_SWIPE:
                _presenter.onSwipe();
                break;
            case TAP:
                _presenter.onTap();
                break;
            case LONG_PRESS:
                _presenter.onLongPress();
                break;
            case DOUBLE_TAP:
                _presenter.onDoubleTap();
                break;
        }
    }

    @Override
    public void showPhotoActionDialog(PhotoActionDialogViewModel viewModel) {
        _photoActionDialog.show(viewModel);
    }

    @Override
    public void setActionDialogViewModel(PhotoActionDialogViewModel viewModel) {
        _photoActionDialog.updateViewModel(viewModel);
    }

    @Override
    public void hidePhotoActionDialog(PhotoScreenContract.HideFlow hideFlow) {
        _photoActionDialog.hide(hideFlow);
    }

    @Override
    public void setFilterType(FilterType filterType) {
        _photoNavBar.setFilterType(filterType);
    }

    @Override
    public void showUI() {
        _photoView.setSystemUiVisibility(
                View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                        | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                        | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN);

        _handler.removeCallbacks(_uiHider);
        _handler.postDelayed(_uiHider, HIDE_UI_DELAY_MS);

        _photoNavBar.show();
    }

    @Override
    public void hideUI() {
        _photoView.setSystemUiVisibility(
                View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                        | View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                        | View.SYSTEM_UI_FLAG_LOW_PROFILE
                        | View.SYSTEM_UI_FLAG_FULLSCREEN
                        | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                        | View.SYSTEM_UI_FLAG_IMMERSIVE
        );

        _photoNavBar.hide();
    }

    @Override
    public void resetPhotoScale() {
        _photoView.resetScale();
    }

    @Override
    public void loadPhoto(String url, boolean notifyOnError) {
        @Nullable RequestListener<Drawable> listener = null;
        if (notifyOnError) {
            listener = new RequestListener<Drawable>() {
                @Override
                public boolean onLoadFailed(@Nullable GlideException e, Object model, Target<Drawable> target, boolean isFirstResource) {
                    _presenter.onImageLoadFailed();

                    return true;
                }

                @Override
                public boolean onResourceReady(Drawable resource, Object model, Target<Drawable> target, DataSource dataSource, boolean isFirstResource) {
                    return false;
                }
            };
        }

        loadPhoto(url, listener);
    }

    private void loadPhoto(String url, @Nullable RequestListener<Drawable> listener) {
        if (getContext() == null) return;

        GlideApp.with(getContext())
                .load(url)
                .diskCacheStrategy(DiskCacheStrategy.NONE)
                .skipMemoryCache(true)
                .listener(listener)
                .into(new DrawableImageViewTarget(_photoView));
    }

    @Override
    public void setPhotoVisible(boolean visible) {
        _photoView.setVisibility(visible ? View.VISIBLE : View.INVISIBLE);
    }

    @Override
    public void onResume() {
        super.onResume();

        _presenter.onResume();
    }

    @Override
    public void onPause() {
        super.onPause();

        _presenter.onPause();

        _handler.removeCallbacks(_uiHider);
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);

        _presenter.saveState(outState);
    }

    @Override
    public void onDestroyView() {
        _photoActionDialog.onDestroyView();
        _photoView.onDestroyView();

        _presenter.onDestroyView();

        super.onDestroyView();
    }
}

File: app/src/main/java/nl/acidcats/tumblrlikes/ui/screens/photo_screen/PhotoFragment.kt
package nl.acidcats.tumblrlikes.ui.screens.photo_screen

import android.graphics.Point
import android.os.Bundle
import android.os.Handler
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.os.bundleOf
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import com.github.ajalt.timberkt.Timber
import kotlinx.android.synthetic.main.fragment_photo.*
import nl.acidcats.tumblrlikes.R
import nl.acidcats.tumblrlikes.ui.Broadcasts
import nl.acidcats.tumblrlikes.ui.screens.base.BaseFragment
import nl.acidcats.tumblrlikes.ui.screens.photo_screen.widgets.InteractiveImageView.Gesture.*
import nl.acidcats.tumblrlikes.ui.screens.photo_screen.widgets.PhotoNavBar
import nl.acidcats.tumblrlikes.ui.screens.photo_screen.widgets.photooptions.PhotoOptionsContract
import nl.acidcats.tumblrlikes.ui.screens.photo_screen.widgets.photooptions.PhotoOptionsContract.Option.*
import nl.acidcats.tumblrlikes.ui.viewmodels.LikesViewModel
import nl.acidcats.tumblrlikes.ui.viewmodels.LikesViewModel.LoadingState
import nl.acidcats.tumblrlikes.ui.viewmodels.PhotoScreenViewModel
import nl.acidcats.tumblrlikes.ui.viewmodels.PhotoScreenViewModel.SaveState
import nl.acidcats.tumblrlikes.util.DeviceUtil
import nl.acidcats.tumblrlikes.util.permissions.PermissionHelper
import nl.acidcats.tumblrlikes.util.permissions.PermissionListener
import javax.inject.Inject

/**
 * Created on 31/10/2018.
 */
class PhotoFragment : BaseFragment() {

    @Inject
    lateinit var permissionHelper: PermissionHelper
    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory

    private val handler: Handler = Handler()
    private val uiHider: Runnable = Runnable { hideUI() }
    private val isTest: Boolean = DeviceUtil.isEmulator
    private lateinit var photoViewModel: PhotoScreenViewModel
    private lateinit var likesViewModel: LikesViewModel
    private val storagePermissionListener: PermissionListener = { permission, isGranted ->
        if (permission == PermissionHelper.Permission.WRITE_EXTERNAL_STORAGE && isGranted) {
            savePhoto()
        }
    }

    companion object {
        private const val HIDE_UI_DELAY_MS = 2000L
        const val KEY_REFRESH = "Refresh"

        fun newInstance(refreshLikes: Boolean): PhotoFragment {
            val fragment = PhotoFragment()

            fragment.arguments = bundleOf(KEY_REFRESH to refreshLikes)

            return fragment
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? = inflater.inflate(R.layout.fragment_photo, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        photoViewModel = ViewModelProviders.of(activity!!, viewModelFactory).get(PhotoScreenViewModel::class.java)
        photoViewModel.getPhoto().observe(this, Observer { photoView.loadPhoto(it.url, it.fallbackUrl) })
        photoViewModel.getFilterType().observe(this, Observer { photoNavBar.setFilter(it) })
        photoViewModel.getSaveState().observe(this, Observer { onPhotoSaveStateChanged(it) })

        likesViewModel = ViewModelProviders.of(activity!!, viewModelFactory).get(LikesViewModel::class.java)
        likesViewModel.getLoadingState().observe(this, Observer { onLikesLoadingStateChanged(it) })

        initPhotoView()

        initPhotoOptionsView()

        initNavBar()

        permissionHelper.addPermissionListener(storagePermissionListener)

        if (isTest) photoView.alpha = .1f

        checkRefreshLikes()
    }

    private fun initNavBar() {
        photoNavBar.filterTypeSelectedListener = { photoViewModel.storeFilterSelection(it) }
        photoNavBar.navBarListener = object : PhotoNavBar.NavBarListener {
            override fun onRefreshRequested() {
                likesViewModel.refreshLikes(LikesViewModel.RefreshType.MANUAL)
            }

            override fun onSettingsRequested() {
                sendBroadcast(Broadcasts.SETTINGS_REQUEST)
            }
        }
    }

    private fun initPhotoOptionsView() {
        photoOptionsView.initViewModel(photoViewModel, this)

        photoOptionsView.setOptionSelectedListener {
            photoOptionsView.hide(PhotoOptionsContract.HideFlow.ANIMATED)

            when (it) {
                FAVORITE -> photoViewModel.togglePhotoFavorite()
                LIKE -> photoViewModel.togglePhotoLike()
                HIDE -> photoViewModel.hidePhoto()
                SAVE -> checkSavePhoto()
            }
        }
    }

    private fun initPhotoView() {
        val screenSize = Point()
        activity?.windowManager?.defaultDisplay?.getRealSize(screenSize)
        photoView.setScreenSize(screenSize)

        photoView.setGestureListener {
            when (it) {
                SIDE_SWIPE -> photoViewModel.getNextPhoto()
                TAP -> photoOptionsView.show()
                LONG_PRESS -> showUI()
                DOUBLE_TAP -> photoView.toggleScaling()
            }
        }
    }

    private fun checkRefreshLikes() {
        val shouldRefresh = arguments?.getBoolean(KEY_REFRESH) ?: false
        if (shouldRefresh) {
            arguments?.putBoolean(KEY_REFRESH, false)

            likesViewModel.refreshLikes(LikesViewModel.RefreshType.AUTOMATIC)
        }
    }

    private fun onLikesLoadingStateChanged(loadingState: LikesViewModel.LoadingState) {
        Timber.d { "onLikesLoadingStateChanged: $loadingState" }

        when (loadingState) {
            is LoadingState.IDLE -> enableRefreshButton(true)
            is LoadingState.LOADING -> enableRefreshButton(false)
            is LoadingState.SUCCESS -> onLikesLoadSuccess()
            is LoadingState.ERROR -> onLikesLoadError()
        }
    }

    private fun enableRefreshButton(enabled: Boolean) {
        photoNavBar.enableRefreshButton(enabled)
    }

    private fun onLikesLoadError() {
        showRefreshCompleteToast(false)

        likesViewModel.resetLoadingState()
    }

    private fun onLikesLoadSuccess() {
        showRefreshCompleteToast(true, likesViewModel.getLoadedLikesCount().value)

        likesViewModel.resetLoadingState()
    }

    private fun showRefreshCompleteToast(success: Boolean, photoCount: Int? = 0) {
        val message = if (success) {
            if (photoCount == 0) {
                getString(R.string.refresh_success_no_new_photos)
            } else {
                getString(R.string.refresh_success, photoCount.toString())
            }
        } else {
            getString(R.string.refresh_error)
        }

        showToast(message)
    }

    override fun onResume() {
        super.onResume()

        hideUI()

        photoViewModel.startPhotoView()
    }

    override fun onPause() {
        super.onPause()

        handler.removeCallbacks(uiHider)

        photoViewModel.endPhotoView()
    }

    private fun showUI() {
        handler.removeCallbacks(uiHider)
        handler.postDelayed(uiHider, HIDE_UI_DELAY_MS)

        photoNavBar.show()
    }

    private fun hideUI() {
        photoView.systemUiVisibility = (
                View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                        or View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                        or View.SYSTEM_UI_FLAG_LOW_PROFILE
                        or View.SYSTEM_UI_FLAG_FULLSCREEN
                        or View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                        or View.SYSTEM_UI_FLAG_IMMERSIVE
                )

        photoNavBar.hide()
    }

    private fun checkSavePhoto() {
        if (context == null || activity == null) return

        if (permissionHelper.hasPermission(context!!, PermissionHelper.Permission.WRITE_EXTERNAL_STORAGE)) {
            savePhoto()
        } else {
            permissionHelper.requestPermission(activity!!, PermissionHelper.Permission.WRITE_EXTERNAL_STORAGE, "")
        }
    }

    private fun savePhoto() {
        photoViewModel.saveBitmap(photoView.getBitmapSnapshot())
    }

    private fun onPhotoSaveStateChanged(saveState: PhotoScreenViewModel.SaveState) {
        Timber.d { "onPhotoSaveStateChanged: $saveState" }

        when (saveState) {
            is SaveState.SUCCESS -> onPhotoSaved(saveState)
            is SaveState.ERROR -> onPhotoSaveError()
        }
    }

    private fun onPhotoSaveError() {
        showToast(getString(R.string.photo_save_error))

        photoViewModel.resetSaveState()
    }

    private fun onPhotoSaved(saveState: PhotoScreenViewModel.SaveState.SUCCESS) {
        showToast(getString(R.string.photo_saved, saveState.filename))

        photoViewModel.resetSaveState()
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        permissionHelper.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    override fun onDestroyView() {
        permissionHelper.removePermissionListener(storagePermissionListener)

        photoOptionsView.onDestroyView()
        photoView.onDestroyView()
        photoNavBar.onDestroyView()

        super.onDestroyView()
    }
}